### PR TITLE
[feat] Redesign job polling mechanism in the framework

### LINKED ISSUE #443: Aggregate job status polls into a single backend command
It would be great if we could poll multiple pending tests using a single backend command such as follows:

```
sacct -j job1,job2,job3
squeue -j job1,job2,job3
```
This optimization will significantly drop the SLURM database hit rates and we could even allow us to have smaller polling intervals, because we could poll multiple job statuses at the same time.

However, the challenging part for this is designing correctly this concept, because it crosses the internal `Job` API. The framework sees jobs as individual entities and does not know how the submission or polling happens. It *can* collect job IDs or jobs, but it *cannot* issue an aggregate `sacct` or `squeue` command, because this is job-scheduler specific. So a new entity must lie in-between to abstract this aggregate operation.

### LINKED ISSUE #1290: Align the internal pipeline of serial and asynchronous policies
Serial and asynchronous policy have a slightly different internal pipeline, that leads to code duplication sometimes.

In the serial policy the `RegressionTask` stages are:
- ...
- `RegressionTask.compile_wait`
- `RegressionTask.run`
- `RegressionTask.wait`, which calls `RegressionTest.wait`. This is a blocking call and periodically checks itself to see when the job is finished, it is polling for the job through the scheduler.
- `RegressionTask.sanity`
- ...

In the asynchronous one the stages are:
- ...
- `RegressionTask.compile_wait`
- `RegressionTask.run`
- `RegressionTask.poll`, a non blocking call that calls `RegressionTest.poll`. When the job is marked as `finished` from the scheduler it will call `RegressionTask.wait`.
- `RegressionTask.wait` (calls the `RegressionTest.wait` as before but the polling here is unnecessary)
- `RegressionTask.sanity`
- ...

On the serial policy, the polling of the job is done through the blocking `wait` method of the scheduler, while on the async one the polling is implemented on the policy itself, by calling the `finished` method of the scheduler. @rsarm had mentioned it first in one of the standups but I came across it and I also think we should change it. I think we should move the "polling" in the scheduler `wait`, from the scheduler to the serial policy and poll instead with the `finished` method like we do on the async policy.

### LINKED ISSUE #1494: `test_submit_timelimit` unit test for the local scheduler fails on Debian
```
_________________________________________________________________________________________ test_submit_timelimit[local] _________________________________________________________________________________________

minimal_job = <reframe.core.schedulers.Job object at 0x7f4190ce6fd0>, local_only = None

    def test_submit_timelimit(minimal_job, local_only):
        minimal_job.time_limit = '2s'
        prepare_job(minimal_job, 'sleep 10')
        t_job = datetime.now()
        minimal_job.submit()
        assert minimal_job.jobid is not None
        minimal_job.wait()
        t_job = datetime.now() - t_job
>       assert t_job.total_seconds() >= 2
E       assert 1.971438 >= 2
E        +  where 1.971438 = <built-in method total_seconds of datetime.timedelta object at 0x7f4190d045a8>()
E        +    where <built-in method total_seconds of datetime.timedelta object at 0x7f4190d045a8> = datetime.timedelta(seconds=1, microseconds=971438).total_seconds

unittests/test_schedulers.py:302: AssertionError
=========================================================================================== short test summary info ============================================================================================
FAILED unittests/test_schedulers.py::test_submit_timelimit[local] - assert 1.971438 >= 2
```

The failure seems to be occasional.

### PR DESCRIPTION
**Pull request by ekouts**
Main idea of this PR:

 - the jobs' state is updated in `poll(*jobs)` instead of `_update_state`. Job-specific errors are not raised immediately but their delivery is deferred until the job status is explicitly queried by its `finished()` method. 
 - the scheduler object doesn't have any information for individual jobs anymore
 - each partition has one common scheduler for all the tests, and the policies should have access to the scheduler objects
 - the former `RegressionTask.poll()`, now renamed to `run_complete()`, checks only the current state and does not trigger any backend command to the scheduler anymore. It also raises an exception for any job-specific error encountered during the polling performed by the scheduler.

Fixes #443 
Fixes #1290 
Fixes #1494

### CODE REVIEW

File: reframe/core/pipeline.py
Comment (vkarak) [inline reframe/core/pipeline.py:1044]: This is a breaking change of the interface, despite giving it a default value. Users override `setup()` with the old signature, and call `super().setup(partition, environ, **job_opts)`, so your `sched` argument will be missed. You should sneak it instead in `job_opts` passing it as a "special" reframe argument, e.g., `_rfm_sched_type`. Then you look for that in `job_opts` inside `setup()`.

File: reframe/core/schedulers/torque.py
Comment (ekouts) [reply reframe/core/schedulers/torque.py:63]: When _all_ the jobs are unknown. In the case that some are valid and some unknown the return code is 0, but it returns information only for the valid ones.

Comment (ekouts) [reply reframe/core/schedulers/torque.py:63]: At least this is what I saw with the dom wrappers. I don't know if the behavior is different in general.

Comment (vkarak) [reply reframe/core/schedulers/torque.py:63]: Add a comment in the code explaining this. It will help understanding.

File: reframe/frontend/executors/policies.py
Comment (ekouts) [reply reframe/frontend/executors/policies.py:226]: Sorry, you are right. I didn't remove it from the constructor but I don't use it anymore. I use `len(self._running_tasks[partname])` when I the number of running per partition.

File: reframe/frontend/executors/policies.py
Comment (ekouts) [reply reframe/frontend/executors/policies.py:380]: I split this because I didn't fit. I will rewrite like this:
```
sched.poll_jobs(
    task.check.job for task in self._running_tasks[partname]
)
```

Comment (vkarak) [reply reframe/frontend/executors/policies.py:380]: The difference is that in the first case you create a list, in the second just a generator.

File: reframe/frontend/executors/policies.py
Comment (ekouts) [reply reframe/frontend/executors/policies.py:464]: To do this I would need to run many long testcases, right? I can get the current polling rate from jenkins, from the reframe.log. Wouldn't it be better if we leave it as it is now and see how many polls are performed after the pr is merged? Because for a few short jobs it wouldn't make a difference, right?

File: reframe/core/schedulers/pbs.py
Comment (ekouts) [reply reframe/core/schedulers/pbs.py:54]: If two jobs are "equal" based only on the jobid, then two jobs from different schedulers could look like they are equal. I can't think of any place that we would use this and compare jobs from different schedulers, I am just making sure that it's fine.

File: reframe/frontend/executors/policies.py
Comment (vkarak) [reply reframe/frontend/executors/policies.py:463]: Now, it's much easier to estimate the poll rate, because it does not depend any more on the number of running tasks. If we keep the poll rate function as of now, we will need to refactor it, because I've seen that you just copied it in two different places.

File: reframe/core/schedulers/pbs.py
Comment (vkarak) [reply reframe/core/schedulers/pbs.py:54]: But you don't mix them, do you? Every scheduler keeps its own jobs.

File: reframe/core/schedulers/__init__.py
Comment (ekouts) [reply reframe/core/schedulers/__init__.py:101]: > The poll_jobs() above should not be abstract, but it should provide the naive implementation, which will go over the jobs list and call poll() on each of the jobs.

The naive implementation will be in the abstract class `JobScheduler`? In cases like slurm where we have the custom aggregate `poll_jobs()` will we have the following?
```
def poll_job(self, job):
    self.poll_jobs([job])
```
Otherwise we will have to have a lot of duplicate code. I made the aggregate `poll_jobs` an abstract method because this is the one that will be called from the policies. I think `poll_job` will be only used from unittests

Comment (vkarak) [reply reframe/core/schedulers/__init__.py:101]: Why don't we define it as `def poll(*jobs)` and have the best of both worlds? You would be able to call it either with a single job or with a list of jobs as `poll(single_job)` or `poll(*multiple_jobs)`.

File: reframe/core/schedulers/torque.py
Comment (vkarak) [inline reframe/core/schedulers/torque.py:58]:
```suggestion
        # If qstat cannot find any of the job IDs, it will return 153.
        # Otherwise, it will return 0 and print information
        # only for the jobs it could find.
```

File: reframe/core/schedulers/slurm.py
Comment (ekouts) [reply reframe/core/schedulers/slurm.py:526]: I thought the `maximum_pending_time` is an attribute of one job. Should the `poll()` method throw an exception, even though only one job has exceeded the maximum pending time?

Comment (vkarak) [reply reframe/core/schedulers/slurm.py:526]: I would say yes. It can through in any case, whereas the `finished()` should just be a state query and should not throw.

Comment (ekouts) [reply reframe/core/schedulers/slurm.py:526]: `finished` was already throwing an error, before the changes. Maybe I misunderstood, which `poll` do you mean? The `poll_jobs` method, that updates the states of all the jobs?

File: reframe/core/schedulers/local.py
Comment (ekouts) [reply reframe/core/schedulers/local.py:181]: `job.jobid` is `None` in `test_kbd_interrupt_in_wait_with_limited_concurrency`

File: reframe/core/schedulers/slurm.py
Comment (vkarak) [reply reframe/core/schedulers/slurm.py:526]: `finished()` in the past was performance what `poll_jobs()` is supposed to do now. `finished()` can now be very lightweight and just report the job state that `poll_jobs()` has updated.

File: reframe/core/schedulers/local.py
Comment (vkarak) [reply reframe/core/schedulers/local.py]: Yes, but what's the scenario there? Why is it `None`? I thought the schedulers kept only valid jobs.

File: reframe/core/schedulers/torque.py
Comment (ekouts) [reply reframe/core/schedulers/torque.py:47]: I get an error in this line if I change it:
```
completed = os_ext.run_command(f"qstat -f {' '.join(jobids)}")
```
```
TypeError: sequence item 0: expected str instance, int found
```

File: reframe/core/schedulers/__init__.py
Comment (vkarak) [inline reframe/core/schedulers/__init__.py:97]:
```suggestion
        :arg jobs: The job descriptors to poll.
```

**Comment by vkarak** [comment]
Some additional comments to this PR:

- The `guess_num_tasks()` should be part of the `JobScheduler` interface (perhaps we can postpone this to address it in a different issue).
- Currently, there are two places that we sleep between polls: (a) inside `JobScheduler.wait()` and inside the busy loop of the asynchronous execution policy. Since everything about job state update comes down to polling, I believe it would be better to implement this sleep inside poll jobs as a centralized poll control mechanism, as suggested by #211. `poll()` could keep track of the polling rate and throttle the polling to meet a certain rate, or it could simply cycle over a set of sleep times.
- One thing that is not very nice is that we keep lots of dictionaries inside, which essentially hold Job properties, e.g., `_job_submit_time`, `_is_cancelling` etc. Some of those properties we could move them to `Job` as public properties, which the scheduler would set and others could be set on-demand by the different backends, e.g., the `_pbs_server` one. If we follow this approach, we could also add a `kind` argument to the `job` that the different backends would set (`job.kind = 'slurm'`); this is just for completeness. This approach would allow us to remove all these dictionaries and perhaps remove the `__eq__` and `__hash__` function from `Job`, although the don't hurt.

File: reframe/core/schedulers/slurm.py
Comment (ekouts) [reply reframe/core/schedulers/slurm.py:392]: I think `finditer` is better because we can still use the named groups. `findall` returns a list of tuples.

Comment (ekouts) [reply reframe/core/schedulers/slurm.py:392]: Same as before, I think `list(re.finditer(...` is better

File: reframe/core/schedulers/slurm.py
Comment (ekouts) [reply reframe/core/schedulers/slurm.py:583]: Same as before, I think `list(re.finditer(...` is better

File: reframe/frontend/executors/policies.py
Comment (ekouts) [reply reframe/frontend/executors/policies.py:442]: I don't understand the change here. We need the name of the partition for the `self._max_jobs` dictionary. We don't want just the tasks

File: reframe/core/schedulers/slurm.py
Comment (vkarak) [reply reframe/core/schedulers/slurm.py:392]: Can we add a comment then, as of why we are doing this? Because I bet it was for this reason and we forgot...

File: reframe/frontend/executors/policies.py
Comment (vkarak) [reply reframe/frontend/executors/policies.py:398]: I've just looked at the diff and missed the rest of the loop. You are right, but I would write it as follows nonetheless:

```suggestion
        for partname, tasks in self._running_tasks.items():
            num_tasks = len(tasks)
```

File: reframe/core/schedulers/slurm.py
Comment (vkarak) [inline reframe/core/schedulers/slurm.py:131]: Keep the comment simple so as to fit in a single line. Also don't use formatting characters for such code comments.

```suggestion
        # We need the match objects, so we have to use finditer()
```

File: reframe/core/schedulers/torque.py
Comment (vkarak) [reply reframe/core/schedulers/torque.py:47]: Ok, I suggest the following then:

```python
if not jobs:
    return

completed = os_ext.run_command(f"qstat -f {' '.join(str(job.jobid) for job in jobs)}")
```

File: reframe/frontend/executors/policies.py
Comment (vkarak) [inline reframe/frontend/executors/policies.py:83]: I think the equivalent to what we have now is the following:

```suggestion
            sleeptime = itertools.cycle(range(1, 10))
```

But we should make this a configuration parameter in any case I think.

File: reframe/frontend/executors/policies.py
Comment (vkarak) [inline reframe/frontend/executors/policies.py:374]:
```suggestion
        for task in itertools.chain(*self._running_tasks.values()):
```

And if `abort()` removes it from the list, then wrap the above inside a `list()`.

**Comment by ekouts** [comment]
@jenkins-cscs retry daint

**Comment by vkarak** [comment]
I have fixed the polling mechanism and fined tuned it. The only remaining thing for this PR now is to revisit the use of `task.poll()`.

**Comment by ekouts** [comment]
> I have fixed the polling mechanism and fined tuned it. The only remaining thing for this PR now is to revisit the use of `task.poll()`.

Would it make sense to rename to `task.update_status()`?

**Comment by vkarak** [comment]
@jenkins-cscs retry daint

**Review by vkarak — APPROVED** [review]

**Review by vkarak — CHANGES_REQUESTED** [review]
Still not ready... It crashes with compile-only regression tests. For example:

```
./bin/reframe --report-file=report.json --prefix=$SCRATCH/rfm-stage/ -C config/cscs.py -c cscs-checks/compile/haswell_fma_check.py -R -t production -t craype -r
```

```
./bin/reframe: unexpected error: 'NoneType' object has no attribute 'scheduler'
Traceback (most recent call last):
  File "/users/karakasv/Devel/reframe/reframe/frontend/cli.py", line 718, in main
    runner.runall(testcases)
  File "/users/karakasv/Devel/reframe/reframe/frontend/executors/__init__.py", line 375, in runall
    self._runall(testcases)
  File "/users/karakasv/Devel/reframe/reframe/frontend/executors/__init__.py", line 428, in _runall
    self._policy.runcase(t)
  File "/users/karakasv/Devel/reframe/reframe/frontend/executors/policies.py", line 315, in runcase
    if not self._setup_task(task):
  File "/users/karakasv/Devel/reframe/reframe/frontend/executors/policies.py", line 284, in _setup_task
    sched_options=self.sched_options)
  File "/users/karakasv/Devel/reframe/reframe/frontend/executors/__init__.py", line 258, in setup
    self._notify_listeners('on_task_setup')
  File "/users/karakasv/Devel/reframe/reframe/frontend/executors/__init__.py", line 221, in _notify_listeners
    callback(self)
  File "/users/karakasv/Devel/reframe/reframe/frontend/executors/policies.py", line 243, in on_task_setup
    self.schedulers.setdefault(partname, task.check.job.scheduler)
AttributeError: 'NoneType' object has no attribute 'scheduler'
```

This is expected, because `job` is never set for this type of tests. The problem is that this problem is not caught by the unit tests.

**Review by vkarak — APPROVED** [review]
It should be fine now.

**Comment by vkarak** [comment]
@jenkins-cscs retry dom

**Review by ekouts — COMMENTED** [review]
I had a look at the latest changes and it looks good to me

**Comment by vkarak** [comment]
There is still a small bug that I have fixed. I am waiting for an extended run to finish, push the fix and merge it.

**Comment by vkarak** [comment]
Final bug fixed. After the CI passes, I think that this PR is finally good to go!

**Pull request merged**
